pressure-vessel/verify.1.md
---
title: pv-verify
section: 1
...

<!-- This document:
Copyright © 2023 Collabora Ltd.
SPDX-License-Identifier: MIT
-->

# NAME

pv-verify - verify integrity of a pressure-vessel-based container

# SYNOPSIS

**pv-verify**
[*OPTIONS*]
[**--**]
[*DIRECTORY*]

**pv-verify --version**

# DESCRIPTION

**pv-verify** checks a pressure-vessel-based container runtime against
a manifest written with a subset of **mtree**(5) syntax.

If no *DIRECTORY* is given, the default is to try to discover the runtime
that contains pressure-vessel.

# OPTIONS

`--minimized-runtime`
:   Assume the *DIRECTORY* is a Steam Linux Runtime runtime in minimized
    form (removing empty files, empty directories, symbolic links and so
    on). This is normally the `*_platform_*/files` directory.

`--mtree`
:   Use a non-default filename for the **mtree**(5) manifest.
    It must currently be compressed with **gzip**(1) or compatible.
    The default is normally `mtree.txt.gz` in the *DIRECTORY*.
    If `--minimized-runtime` is used, the default is `../usr-mtree.txt.gz`
    relative to the *DIRECTORY*.

`--quiet`
:   Don't show informational messages if the runtime is successfully verified,
    just exit with status 0.

`--verbose`
:   Be more verbose.

`--version`
:   Print the version number and exit.

# ENVIRONMENT

`SRT_LOG`
:   A sequence of tokens separated by colons, spaces or commas
    affecting how output is recorded. See source code for details.

# OUTPUT

Unstructured diagnostic messages are output on standard error.

# EXIT STATUS

0
:   The runtime matches the manifest

Anything else
:   The runtime does not match the manifest, or an error occurred

# EXAMPLE

    $ steam steam://install/1070560     # Steam Linux Runtime 1.0 (scout)
    $ cd ~/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier
    $ ./pressure-vessel/bin/pv-verify
    $ ./pressure-vessel/bin/pv-verify ../SteamLinuxRuntime

<!-- vim:set sw=4 sts=4 et: -->


debian/changelog
steam-runtime-tools (0.20231127.0) UNRELEASED; urgency=medium

  * pressure-vessel:
    - Log a warning if we can't set up /etc/timezone
    - If emitting debug messages, include the full contents of the new
      ld.so.cache (steamrt/tasks#357, steam-runtime#630)
    - Make SRT_LOG=debug activate extra output.
      Previously, only --verbose or PRESSURE_VESSEL_VERBOSE would have
      activated this.
    - Better compatibility with very old CPUs, especially when using VDPAU,
      if the host OS has a supported expansion of ${LIB}
      + Affected Intel CPUs are older than "Haswell" (2013)
      + Affected AMD CPUs are older than "Excavator" (2015)
      + Systems where this works are those that use multiarch
        (Debian/Ubuntu or Flatpak), FHS multilib (Red Hat, etc.) or
        Arch Linux's multilib variant. Other systems where ${LIB} expands
        to distinct values for 32- and 64-bit x86 can easily be supported
        if their users open an issue with the details.
    - Better compatibility with FEX-Emu, especially when using VDPAU
      (steamrt/tasks#349)
    - pv-verify: New tool to verify a directory against a mtree(5) manifest
    - Refactoring
  * Diagnostic tools:
    - Improve performance of steam-runtime-system-info, especially when
      running under FEX-Emu emulation (steam-runtime-tools#88)
    - Stop relying on ability to run timeout(1), which might improve
      compatibility with some sandboxed or limited environments
      (steam-runtime-tools#88)
    - Resync wflinfo implementation with waffle 1.8.0
  * Steam Linux Runtime depots:
    - In usr-mtree.txt.gz, only put permissions in mode=, not all of st_mode,
      matching bsdtar(1) and mtree(8)
    - In usr-mtree.txt.gz, truncate fractional seconds to ".0" if zero
    - Add a top-level mtree manifest for the depot as a whole.
      In SLR 2.0 (soldier) and 3.0 (sniper), after pressure-vessel has been
      updated to a suitable version, it will be possible to verify the
      depot against its manifest by running a command like:
        /path/to/SteamLinuxRuntime_soldier/pressure-vessel/bin/pv-verify
      In SLR 1.0 (scout) the equivalent is:
        /path/to/SteamLinuxRuntime_soldier/pressure-vessel/bin/pv-verify \
            path/to/SteamLinuxRuntime
  * Library code:
    - Refactor helper subprocess launching
    - Make execution environment for helper subprocesses more consistent
  * Tests:
    - Make the pv-adverb test coverage more realistic
    - When using valgrind, don't show reachable objects by default
    - Fix a minor memory leak in check-xdg-portal (not significant during
      production use since this is not a long-lived process)
    - Stop diagnosing memory leaks in wflinfo (not significant during
      production use since this is not a long-lived process)

 -- Simon McVittie <smcv@collabora.com>  Mon, 27 Nov 2023 13:07:41 +0000

steam-runtime-tools (0.20231107.1) steamrt; urgency=medium

  * Fix internal version information

 -- Simon McVittie <smcv@collabora.com>  Tue, 07 Nov 2023 14:03:21 +0000

steam-runtime-tools (0.20231107.0) steamrt; urgency=medium

  * All CLI tools: Consistently unblock and unignore signals during startup
    (steamrt/tasks#354)
  * system-info:
    - Document JSON format for virtualization
    - Add an option to skip checking libraries
    - Add an option to skip graphics smoke-tests
    - Use the common log handler, so SRT_LOG takes effect
    - steam-runtime-system-info --verbose now outputs all fields from
      os-release(5)
  * pressure-vessel: Refactoring
  * urlopen: Use steam://openurl_external to open http, https URLs
    (steamrt/tasks#80)
  * Library code:
    - SrtSystemInfo: Avoid using fd-relative I/O for the real root, making
      several diagnostics more accurate when running under FEX-Emu:
      * DRI, VA-API and VDPAU modules (steamrt/tasks#347)
      * Vulkan layers
      * os-release(5) (steamrt/tasks#342)
      * ld.so (steamrt/tasks#348)
      * /overrides
    - SrtSystemInfo: Make more use of fd-relative I/O when acting on a
      sysroot, avoiding being misled by absolute symbolic links
    - SrtSystemInfo: Break reference cycles in dispose() if necessary
    - SrtSystemInfo: Fix assertion failure if failing to get container
      information
    - SrtSystemInfo: Keep better track of whether opening a sysroot failed
    - SrtVirtualizationInfo: Add missing getter for host-machine property
    - SrtVirtualizationInfo: Add os-release(5) from the host system under
      FEX-Emu
    - Collect os-release(5) path, canonicalized path, and any diagnostic
      messages about errors parsing it, not just its contents
    - Improve test coverage
    - Improve build robustness in scout
    - Refactoring
  * Tests:
    - Move non-pressure-vessel-specific tests to tests/
    - Fix test failure with a custom value for LC_ADDRESS
    - Remove unused imports

 -- Simon McVittie <smcv@collabora.com>  Tue, 07 Nov 2023 12:10:59 +0000

steam-runtime-tools (0.20231013.1) steamrt; urgency=medium

  * launcher-service: Unblock and unignore all signals during startup,
    fixing detection of subprocesses exiting when launched by Steam
    (steamrt/tasks#354, ValveSoftware/SteamVR-for-Linux#618)

 -- Simon McVittie <smcv@collabora.com>  Tue, 24 Oct 2023 19:47:07 +0100

steam-runtime-tools (0.20231013.0) steamrt; urgency=medium

  * pressure-vessel:
    - Add more entries to the list of NVIDIA driver libraries,
      notably libnvidia-gpucomp. Thanks to Daniel Dadap (steamrt/tasks#340)
    - pv-runtime: Duplicate symlinks in /etc, /var into the FEX-Emu
      interpreter root, fixing incorrect /etc/os-release under FEX-Emu
      (steamrt/tasks#342)
    - pv-adverb: Reduce environment variables, etc. to debug level.
      This reduces noise when redirecting to the systemd Journal.
    - pv-runtime: Use a closer match for 99-pulseaudio-default.conf
      when the host system uses PulseAudio.
      Thanks to Nozomi Miyamori (steam-runtime#623)

pressure-vessel/verify.c
/* pressure-vessel-verify — verify SteamLinuxRuntime_* against a manifest
 *
 * Copyright © 2023 Collabora Ltd.
 * SPDX-License-Identifier: LGPL-2.1-or-later
 */

#include "config.h"

#include <locale.h>
#include <sysexits.h>
#include <unistd.h>

#include "libglnx.h"

#include "steam-runtime-tools/glib-backports-internal.h"
#include "steam-runtime-tools/log-internal.h"
#include "steam-runtime-tools/utils-internal.h"

#include "mtree.h"

static gboolean opt_minimized_runtime = FALSE;
static gchar *opt_mtree = NULL;
static gboolean opt_quiet = FALSE;
static gboolean opt_verbose = FALSE;
static gboolean opt_version = FALSE;

static GOptionEntry options[] =
{
  { "minimized-runtime", '\0',
    G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &opt_minimized_runtime,
    "Verify a minimized runtime.", NULL },
  { "mtree", '\0',
    G_OPTION_FLAG_NONE, G_OPTION_ARG_FILENAME, &opt_mtree,
    "Verify a manifest other than the default filename.", NULL },
  { "quiet", '\0',
    G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &opt_quiet,
    "Be less verbose.", NULL },
  { "verbose", '\0',
    G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &opt_verbose,
    "Be more verbose.", NULL },
  { "version", '\0',
    G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &opt_version,
    "Print version number and exit.", NULL },
  { NULL }
};

static void
no_log_handler (const gchar *log_domain,
                GLogLevelFlags log_level,
                const gchar *message,
                gpointer user_data)
{
}

static gboolean
run (int argc,
     char **argv,
     GError **error)
{
  g_autoptr(GOptionContext) context = NULL;
  g_autofree gchar *mtree = NULL;
  g_autofree gchar *top = NULL;
  glnx_autofd int top_fd = -1;
  PvMtreeApplyFlags flags = PV_MTREE_APPLY_FLAGS_GZIP;

  setlocale (LC_ALL, "");

  if (!_srt_util_set_glib_log_handler ("pv-verify",
                                       G_LOG_DOMAIN,
                                       (SRT_LOG_FLAGS_OPTIONALLY_JOURNAL
                                        | SRT_LOG_FLAGS_DIVERT_STDOUT),
                                       NULL, NULL, error))
    return FALSE;

  context = g_option_context_new ("DIRECTORY\n"
                                  "DIRECTORY is SteamLinuxRuntime_sniper or similar.\n");
  g_option_context_add_main_entries (context, options, NULL);

  if (!g_option_context_parse (context, &argc, &argv, error))
    return FALSE;

  if (opt_version)
    {
      g_print ("%s:\n"
               " Package: steam-runtime-tools\n"
               " Version: %s\n",
               g_get_prgname (), VERSION);
      return TRUE;
    }

  if (!_srt_util_set_glib_log_handler (NULL, G_LOG_DOMAIN,
                                       (SRT_LOG_FLAGS_OPTIONALLY_JOURNAL |
                                        (opt_verbose ? SRT_LOG_FLAGS_DEBUG : 0)),
                                       NULL, NULL, error))
    return FALSE;

  if (opt_quiet)
    g_log_set_handler (G_LOG_DOMAIN,
                       G_LOG_LEVEL_MESSAGE,
                       no_log_handler,
                       NULL);

  _srt_unblock_signals ();
  _srt_setenv_disable_gio_modules ();

  if (argc >= 2 && strcmp (argv[1], "--") == 0)
    {
      argv++;
      argc--;
    }

  if (argc > 2)
    {
      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED,
                   "Usage: %s [OPTIONS] [DIRECTORY]", g_get_prgname ());
      return FALSE;
    }
  else if (argc == 2)
    {
      top = g_strdup (argv[1]);
    }
  else
    {
      const char *prefix = _srt_find_myself (NULL, error);

      if (prefix == NULL)
        return FALSE;

      top = g_path_get_dirname (prefix);
    }

  if (!glnx_opendirat (AT_FDCWD, top, TRUE, &top_fd, error))
    return FALSE;

  if (opt_mtree)
    mtree = g_strdup (opt_mtree);
  else if (opt_minimized_runtime)
    mtree = g_build_filename (top, "../usr-mtree.txt.gz", NULL);
  else
    mtree = g_build_filename (top, "mtree.txt.gz", NULL);

  if (opt_minimized_runtime)
    flags |= PV_MTREE_APPLY_FLAGS_MINIMIZED_RUNTIME;

  if (!pv_mtree_verify (mtree, top, top_fd, flags, error))
    return FALSE;

  return TRUE;
}

int
main (int argc,
      char *argv[])
{
  g_autoptr(GError) local_error = NULL;
  GError **error = &local_error;
  int ret = 0;

  if (run (argc, argv, error))
    goto out;

  g_return_val_if_fail (local_error != NULL, EX_SOFTWARE);
  _srt_log_failure ("%s", local_error->message);

  if (local_error->domain == G_OPTION_ERROR)
    ret = EX_USAGE;
  else
    ret = 1;

out:
  g_free (opt_mtree);
  return ret;
}
